File: examples/calculator-java-hexagonal-cqrs/src/main/java/com/oopforge/example/calculator/adapter/persistence/InMemoryHistoryQueryRepository.java
package com.oopforge.example.calculator.adapter.persistence;

import com.oopforge.example.calculator.application.query.HistoryQueryRepository;
import com.oopforge.example.calculator.application.query.HistorySummary;
import com.oopforge.example.calculator.domain.CalculationId;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

/**
 * Read adapter: serves projections from the store, most-recent first.
 */
public class InMemoryHistoryQueryRepository implements HistoryQueryRepository {

    private final CalculationStore store;

    public InMemoryHistoryQueryRepository(CalculationStore store) {
        this.store = store;
    }

    @Override
    public List<HistorySummary> listRecent(int limit) {
        if (limit <= 0) {
            throw new IllegalArgumentException("limit must be positive");
        }
        List<HistorySummary> all = store.historySnapshot();
        int from = Math.max(0, all.size() - limit);
        List<HistorySummary> tail = new ArrayList<>(all.subList(from, all.size()));
        Collections.reverse(tail);
        return List.copyOf(tail);
    }

    @Override
    public Optional<HistorySummary> findById(CalculationId calculationId) {
        String target = calculationId.value().toString();
        List<HistorySummary> all = store.historySnapshot();
        for (int i = all.size() - 1; i >= 0; i--) {
            if (all.get(i).calculationId().equals(target)) {
                return Optional.of(all.get(i));
            }
        }
        return Optional.empty();
    }
}

File: examples/calculator-python-hexagonal-cqrs/app/infrastructure/repositories/in_memory_history_query_repository.py
from app.application.query.history_summary import HistorySummary
from app.domain.calculation.value import CalculationId
from app.infrastructure.repositories.store import CalculationStore


class InMemoryHistoryQueryRepository:
    """Read adapter: serves projections from the store."""

    def __init__(self, store: CalculationStore) -> None:
        self._store = store

    def list_recent(self, limit: int) -> list[HistorySummary]:
        if limit <= 0:
            raise ValueError("limit must be positive")
        return list(reversed(self._store.history[-limit:]))

    def find_by_id(self, calculation_id: CalculationId) -> HistorySummary | None:
        for entry in reversed(self._store.history):
            if entry.calculation_id == str(calculation_id.value):
                return entry
        return None
